Groovy parser fails to parse multiline strings

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
I am using

- OpenRewrite main branch (v8.79.6)

## How are you running OpenRewrite?

I am running a new unit test in `rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java`:
```
diff --git a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
index 81ee71d23..6f2428e28 100644
--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
@@ -337,4 +337,29 @@ def method() {
         );
     }
 
+    @Test
+    void testMultilineString() {
+        rewriteRun(
+          groovy(
+            """
+                String name = \"\"\"
+                    foo
+                \"\"\"
+                """
+          )
+        );
+    }
+    @Test
+    void testMultilineStringInterpolation() {
+        rewriteRun(
+          groovy(
+            """
+                String name = \"\"\"
+                    ${foo}
+                \"\"\"
+                """
+          )
+        );
+    }
+
 }

```

## What is the smallest, simplest way to reproduce the problem?

```groovy
String test = """
 ${foo}
"""
```

## What did you expect to see?

Code parsed successfully

## What did you see instead?

Parser failure

## What is the full stack trace of any errors you encountered?
```

> Configure project :
Inferred project: rewrite, version: 8.80.0-SNAPSHOT
> Task :rewrite-kotlin:checkKotlinGradlePluginConfigurationErrors SKIPPED
> Task :rewrite-java-21:processResources NO-SOURCE
> Task :rewrite-java-21:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-java-test:processResources UP-TO-DATE
> Task :rewrite-java:processResources UP-TO-DATE
> Task :rewrite-groovy:processResources UP-TO-DATE
> Task :rewrite-java-test:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-groovy:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-java:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-java-21:writeManifestProperties
> Task :rewrite-java-test:writeManifestProperties
> Task :rewrite-groovy:writeManifestProperties
> Task :rewrite-java:writeManifestProperties
> Task :rewrite-kotlin:processResources UP-TO-DATE
> Task :rewrite-groovy:processTestResources UP-TO-DATE
> Task :rewrite-test:processResources UP-TO-DATE
> Task :rewrite-kotlin:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-test:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-kotlin:writeManifestProperties
> Task :rewrite-test:writeManifestProperties
> Task :rewrite-java-lombok:unpackAndRenameLombok UP-TO-DATE
> Task :rewrite-java-lombok:processResources UP-TO-DATE
> Task :rewrite-java-lombok:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-java-lombok:writeManifestProperties
> Task :rewrite-core:compileJava UP-TO-DATE
> Task :rewrite-core:processResources UP-TO-DATE
> Task :rewrite-core:classes UP-TO-DATE
> Task :rewrite-core:createPropertiesFileForJar UP-TO-DATE
> Task :rewrite-core:writeManifestProperties
> Task :rewrite-java-lombok:compileJava UP-TO-DATE
> Task :rewrite-core:jar UP-TO-DATE
> Task :rewrite-java-lombok:classes UP-TO-DATE
> Task :rewrite-java-lombok:jar UP-TO-DATE
> Task :rewrite-test:compileJava UP-TO-DATE
> Task :rewrite-test:classes UP-TO-DATE
> Task :rewrite-test:jar UP-TO-DATE
> Task :rewrite-yaml:compileJava UP-TO-DATE
> Task :rewrite-java:compileJava UP-TO-DATE
> Task :rewrite-java:classes UP-TO-DATE
> Task :rewrite-java:shadowJar UP-TO-DATE
> Task :rewrite-kotlin:compileKotlin UP-TO-DATE
> Task :rewrite-kotlin:compileJava UP-TO-DATE
> Task :rewrite-kotlin:classes UP-TO-DATE
> Task :rewrite-kotlin:jar UP-TO-DATE
> Task :rewrite-java-21:compileJava UP-TO-DATE
> Task :rewrite-java-21:classes UP-TO-DATE
> Task :rewrite-java-21:jar UP-TO-DATE
> Task :rewrite-groovy:compileJava UP-TO-DATE
> Task :rewrite-groovy:classes UP-TO-DATE
> Task :rewrite-groovy:jar UP-TO-DATE
> Task :rewrite-java-test:compileJava UP-TO-DATE
> Task :rewrite-java-test:classes UP-TO-DATE
> Task :rewrite-java-test:jar UP-TO-DATE
> Task :rewrite-groovy:compileTestJava UP-TO-DATE
> Task :rewrite-groovy:testClasses UP-TO-DATE

Failed to parse sources or run recipe
java.lang.AssertionError: Failed to parse sources or run recipe
	at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:697)
	at org.openrewrite.test.RewriteTest$$Lambda/0x00007849c0184de8.accept(Unknown Source)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:160)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007849c01cfbd0.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:301)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:135)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
	at org.openrewrite.groovy.GroovyParserTest.testMultilineStringInterpolation(GroovyParserTest.java:354)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x00007849c017c000.invokeVirtual(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x00007849c00e0800.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 28. The surrounding characters in the original source are:
String name = """
    ${foo}~cursor~>
"""
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:223)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:149)
	... 19 more
Caused by: java.lang.IllegalArgumentException: Couldn't find delimiter: """ with fromIndex: 31
	at org.openrewrite.groovy.GroovyParserVisitor.sourceSubstring(GroovyParserVisitor.java:3108)
	at org.openrewrite.groovy.GroovyParserVisitor.access$1800(GroovyParserVisitor.java:81)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitGStringExpression(GroovyParserVisitor.java:1968)
	at org.codehaus.groovy.ast.expr.GStringExpression.visit(GStringExpression.java:53)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.doVisit(GroovyParserVisitor.java:822)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitDeclarationExpression(GroovyParserVisitor.java:1807)
	at org.codehaus.groovy.ast.expr.DeclarationExpression.visit(DeclarationExpression.java:89)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$11(GroovyParserVisitor.java:1847)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor$$Lambda/0x00007849c03089a0.get(Unknown Source)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:878)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1846)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:2736)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:209)
	... 20 more


Failed to parse file.groovy at cursor position 28. The surrounding characters in the original source are:
String name = """
    ${foo}~cursor~>
"""
org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 28. The surrounding characters in the original source are:
String name = """
    ${foo}~cursor~>
"""
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:223)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:149)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007849c01cfbd0.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:301)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:135)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
	at org.openrewrite.groovy.GroovyParserTest.testMultilineStringInterpolation(GroovyParserTest.java:354)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x00007849c017c000.invokeVirtual(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x00007849c00e0800.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.IllegalArgumentException: Couldn't find delimiter: """ with fromIndex: 31
	at org.openrewrite.groovy.GroovyParserVisitor.sourceSubstring(GroovyParserVisitor.java:3108)
	at org.openrewrite.groovy.GroovyParserVisitor.access$1800(GroovyParserVisitor.java:81)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitGStringExpression(GroovyParserVisitor.java:1968)
	at org.codehaus.groovy.ast.expr.GStringExpression.visit(GStringExpression.java:53)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.doVisit(GroovyParserVisitor.java:822)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitDeclarationExpression(GroovyParserVisitor.java:1807)
	at org.codehaus.groovy.ast.expr.DeclarationExpression.visit(DeclarationExpression.java:89)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$11(GroovyParserVisitor.java:1847)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor$$Lambda/0x00007849c03089a0.get(Unknown Source)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:878)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1846)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:2736)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:209)
	... 20 more


Couldn't find delimiter: """ with fromIndex: 31
java.lang.IllegalArgumentException: Couldn't find delimiter: """ with fromIndex: 31
	at org.openrewrite.groovy.GroovyParserVisitor.sourceSubstring(GroovyParserVisitor.java:3108)
	at org.openrewrite.groovy.GroovyParserVisitor.access$1800(GroovyParserVisitor.java:81)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitGStringExpression(GroovyParserVisitor.java:1968)
	at org.codehaus.groovy.ast.expr.GStringExpression.visit(GStringExpression.java:53)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.doVisit(GroovyParserVisitor.java:822)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitDeclarationExpression(GroovyParserVisitor.java:1807)
	at org.codehaus.groovy.ast.expr.DeclarationExpression.visit(DeclarationExpression.java:89)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$11(GroovyParserVisitor.java:1847)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor$$Lambda/0x00007849c03089a0.get(Unknown Source)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:878)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1846)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:2736)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:209)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:149)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007849c01cfbd0.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:301)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:135)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
	at org.openrewrite.groovy.GroovyParserTest.testMultilineStringInterpolation(GroovyParserTest.java:354)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x00007849c017c000.invokeVirtual(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x00007849c00e0800.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


> Task :rewrite-groovy:test FAILED
GroovyParserTest > testMultilineStringInterpolation() FAILED
    java.lang.AssertionError: Failed to parse sources or run recipe
        at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:697)
        at org.openrewrite.test.RewriteTest$$Lambda/0x00007849c0184de8.accept(Unknown Source)
        at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:160)
        at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007849c01cfbd0.apply(Unknown Source)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:301)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:135)
        at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
        at org.openrewrite.groovy.GroovyParserTest.testMultilineStringInterpolation(GroovyParserTest.java:354)
        Caused by:
        org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 28. The surrounding characters in the original source are:
        String name = """
            ${foo}~cursor~>
        """
            at app//org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:223)
            at app//org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:149)
            ... 13 more
            Caused by:
            java.lang.IllegalArgumentException: Couldn't find delimiter: """ with fromIndex: 31
                at org.openrewrite.groovy.GroovyParserVisitor.sourceSubstring(GroovyParserVisitor.java:3108)
                at org.openrewrite.groovy.GroovyParserVisitor.access$1800(GroovyParserVisitor.java:81)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitGStringExpression(GroovyParserVisitor.java:1968)
                at org.codehaus.groovy.ast.expr.GStringExpression.visit(GStringExpression.java:53)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.doVisit(GroovyParserVisitor.java:822)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitDeclarationExpression(GroovyParserVisitor.java:1807)
                at org.codehaus.groovy.ast.expr.DeclarationExpression.visit(DeclarationExpression.java:89)
                at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.lambda$visitExpressionStatement$11(GroovyParserVisitor.java:1847)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor$$Lambda/0x00007849c03089a0.get(Unknown Source)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.labeled(GroovyParserVisitor.java:878)
                at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitExpressionStatement(GroovyParserVisitor.java:1846)
                at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
                at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:2736)
                at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:209)
                ... 14 more
1 test completed, 1 failed
[Incubating] Problems report is available at: file:///home/vladimirp/git/rewrite/build/reports/problems/problems-report.html
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':rewrite-groovy:test'.
> There were failing tests. See the report at: file:///home/vladimirp/git/rewrite/rewrite-groovy/build/reports/tests/test/index.html
Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
For more on this, please refer to https://docs.gradle.org/9.3.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
BUILD FAILED in 20s
45 actionable tasks: 9 executed, 36 up-to-date

```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -1965,7 +1965,7 @@ public class GroovyParserVisitor {
                     }
                 } else if (e instanceof ConstantExpression) {
                     // Get the string literal from the source, so escaping of newlines and the like works out of the box
-                    String value = sourceSubstring(cursor, delimiter.close);
+                    String value = sourceSubstring(cursor, delimiter.close, cursor + e.getText().length());
                     // There could be a closer GString before the end of the closing delimiter, so shorten the string if needs be
                     int indexNextSign = source.indexOf("$", cursor);
                     while (isEscaped(indexNextSign, delimiter)) {
@@ -3099,10 +3099,13 @@ public class GroovyParserVisitor {
 
     /**
      * Returns a string that is a part of this source. The substring begins at the specified beginIndex and extends until delimiter.
+     * The search starts from the fromIndex.
      * The cursor will not be moved.
      */
-    private String sourceSubstring(int beginIndex, String untilDelim) {
-        int fromIndex = Math.max(beginIndex, cursor + untilDelim.length());
+    private String sourceSubstring(int beginIndex, String untilDelim, int fromIndex) {
+        if (fromIndex < beginIndex) {
+            throw new IllegalArgumentException("beginIndex: " + beginIndex + " should be < fromIndex: "+ fromIndex);
+        }
         int endIndex = source.indexOf(untilDelim, fromIndex);
         if (endIndex < 0) {
             throw new IllegalArgumentException(
@@ -3120,6 +3123,15 @@ public class GroovyParserVisitor {
             );
         }
         return source.substring(beginIndex, endIndex);
+    }
+
+    /**
+     * Returns a string that is a part of this source. The substring begins at the specified beginIndex and extends until delimiter.
+     * The cursor will not be moved.
+     */
+    private String sourceSubstring(int beginIndex, String untilDelim) {
+        int fromIndex = Math.max(beginIndex, cursor + 1);
+        return sourceSubstring(beginIndex, untilDelim, fromIndex);
     }
 
     private @Nullable Integer getInsideParenthesesLevel(ASTNode node) {

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -3104,12 +3104,12 @@ public class GroovyParserVisitor {
      */
     private String sourceSubstring(int beginIndex, String untilDelim, int fromIndex) {
         if (fromIndex < beginIndex) {
-            throw new IllegalArgumentException("beginIndex: " + beginIndex + " should be < fromIndex: "+ fromIndex);
+            throw new IllegalArgumentException("beginIndex: " + beginIndex + " should be < fromIndex: " + fromIndex);
         }
         int endIndex = source.indexOf(untilDelim, fromIndex);
         if (endIndex < 0) {
             throw new IllegalArgumentException(
-                "Couldn't find delimiter: " + untilDelim + " with fromIndex: " + fromIndex
+                    "Couldn't find delimiter: " + untilDelim + " with fromIndex: " + fromIndex
             );
         }
         // don't stop if the last char is escaped.
@@ -3119,7 +3119,7 @@ public class GroovyParserVisitor {
         }
         if (endIndex < 0) {
             throw new IllegalArgumentException(
-                "Couldn't find unescaped delimiter: " + untilDelim + " with fromIndex: " + fromIndex
+                    "Couldn't find unescaped delimiter: " + untilDelim + " with fromIndex: " + fromIndex
             );
         }
         return source.substring(beginIndex, endIndex);

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -1948,6 +1948,7 @@ public class GroovyParserVisitor {
                 sortedByPosition.put(pos(e), e);
             }
             List<org.codehaus.groovy.ast.expr.Expression> rawExprs = new ArrayList<>(sortedByPosition.values());
+            boolean hasInterpolation = !gstring.getValues().isEmpty();
             List<J> strings = new ArrayList<>(rawExprs.size());
             for (int i = 0; i < rawExprs.size(); i++) {
                 org.codehaus.groovy.ast.expr.Expression e = rawExprs.get(i);
@@ -1965,15 +1966,9 @@ public class GroovyParserVisitor {
                     }
                 } else if (e instanceof ConstantExpression) {
                     // Get the string literal from the source, so escaping of newlines and the like works out of the box
-                    String value = sourceSubstring(cursor, delimiter.close, cursor + e.getText().length());
-                    // There could be a closer GString before the end of the closing delimiter, so shorten the string if needs be
-                    int indexNextSign = source.indexOf("$", cursor);
-                    while (isEscaped(indexNextSign, delimiter)) {
-                        indexNextSign = source.indexOf("$", indexNextSign + 1);
-                    }
-                    if (indexNextSign != -1 && indexNextSign < (cursor + value.length())) {
-                        value = source.substring(cursor, indexNextSign);
-                    }
+                    String value = hasInterpolation ?
+                            readConstantSegmentBeforeNextInterpolation(delimiter) :
+                            sourceSubstring(cursor, delimiter.close);
                     strings.add(new J.Literal(randomId(), EMPTY, Markers.EMPTY, value, value, null, JavaType.Primitive.String));
                     skip(value);
                 } else {
@@ -3099,17 +3094,14 @@ public class GroovyParserVisitor {
 
     /**
      * Returns a string that is a part of this source. The substring begins at the specified beginIndex and extends until delimiter.
-     * The search starts from the fromIndex.
      * The cursor will not be moved.
      */
-    private String sourceSubstring(int beginIndex, String untilDelim, int fromIndex) {
-        if (fromIndex < beginIndex) {
-            throw new IllegalArgumentException("beginIndex: " + beginIndex + " should be < fromIndex: " + fromIndex);
-        }
+    private String sourceSubstring(int beginIndex, String untilDelim) {
+        int fromIndex = Math.max(beginIndex, cursor + untilDelim.length());
         int endIndex = source.indexOf(untilDelim, fromIndex);
         if (endIndex < 0) {
             throw new IllegalArgumentException(
-                    "Couldn't find delimiter: " + untilDelim + " with fromIndex: " + fromIndex
+                "Couldn't find delimiter: " + untilDelim + " with fromIndex: " + fromIndex
             );
         }
         // don't stop if the last char is escaped.
@@ -3119,19 +3111,42 @@ public class GroovyParserVisitor {
         }
         if (endIndex < 0) {
             throw new IllegalArgumentException(
-                    "Couldn't find unescaped delimiter: " + untilDelim + " with fromIndex: " + fromIndex
+                "Couldn't find unescaped delimiter: " + untilDelim + " with fromIndex: " + fromIndex
             );
         }
         return source.substring(beginIndex, endIndex);
     }
 
     /**
-     * Returns a string that is a part of this source. The substring begins at the specified beginIndex and extends until delimiter.
-     * The cursor will not be moved.
+     * Reads a constant (literal) segment of a GString, starting at the current {@code cursor} and
+     * ending just before either the next interpolation sign ({@code $}) or the closing delimiter.
+     * Unlike {@link #sourceSubstring(int, String)} this does not scan the entire remaining source
+     * for the closing delimiter, which is unreliable for multi-line strings that may contain
+     * embedded quote sequences.
      */
-    private String sourceSubstring(int beginIndex, String untilDelim) {
-        int fromIndex = Math.max(beginIndex, cursor + 1);
-        return sourceSubstring(beginIndex, untilDelim, fromIndex);
+    private String readConstantSegmentBeforeNextInterpolation(Delimiter delimiter) {
+        int indexNextSign = source.indexOf("$", cursor);
+        while (indexNextSign != -1 && isEscaped(indexNextSign, delimiter)) {
+            indexNextSign = source.indexOf("$", indexNextSign + 1);
+        }
+        int indexCloseDelim = source.indexOf(delimiter.close, cursor);
+        while (indexCloseDelim != -1 && isEscaped(indexCloseDelim)) {
+            indexCloseDelim = source.indexOf(delimiter.close, indexCloseDelim + 1);
+        }
+        int endIndex;
+        if (indexNextSign == -1) {
+            endIndex = indexCloseDelim;
+        } else if (indexCloseDelim == -1) {
+            endIndex = indexNextSign;
+        } else {
+            endIndex = Math.min(indexNextSign, indexCloseDelim);
+        }
+        if (endIndex < 0) {
+            throw new IllegalArgumentException(
+                "Couldn't find end of GString constant segment starting at cursor: " + cursor
+            );
+        }
+        return source.substring(cursor, endIndex);
     }
 
     private @Nullable Integer getInsideParenthesesLevel(ASTNode node) {

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
@@ -363,4 +363,55 @@ class GroovyParserTest implements RewriteTest {
         );
     }
 
+    @Test
+    void testMultilineStringInterpolationFollowedByTripleQuote() {
+        rewriteRun(
+          groovy(
+            """
+              String name = \"""${foo}\"""
+              """
+          )
+        );
+    }
+
+    @Test
+    void testMultilineStringDollarVariable() {
+        rewriteRun(
+          groovy(
+            """
+              def foo = "x"
+              String name = \"""
+                  $foo bar
+              \"""
+              """
+          )
+        );
+    }
+
+    @Test
+    void testMultilineStringWithEmbeddedQuotes() {
+        rewriteRun(
+          groovy(
+            """
+              String name = \"""
+                  "quoted" and ${foo}
+              \"""
+              """
+          )
+        );
+    }
+
+    @Test
+    void testMultilineStringWithEscapedDollar() {
+        rewriteRun(
+          groovy(
+            """
+              String name = \"""
+                  \\$literal then ${foo}
+              \"""
+              """
+          )
+        );
+    }
+
 }

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
@@ -342,22 +342,23 @@ class GroovyParserTest implements RewriteTest {
         rewriteRun(
           groovy(
             """
-                String name = \"\"\"
-                    foo
-                \"\"\"
-                """
+              String name = \"""
+                  foo
+              \"""
+              """
           )
         );
     }
+
     @Test
     void testMultilineStringInterpolation() {
         rewriteRun(
           groovy(
             """
-                String name = \"\"\"
-                    ${foo}
-                \"\"\"
-                """
+              String name = \"""
+                  ${foo}
+              \"""
+              """
           )
         );
     }

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/GroovyParserTest.java
@@ -337,4 +337,29 @@ class GroovyParserTest implements RewriteTest {
         );
     }
 
+    @Test
+    void testMultilineString() {
+        rewriteRun(
+          groovy(
+            """
+                String name = \"\"\"
+                    foo
+                \"\"\"
+                """
+          )
+        );
+    }
+    @Test
+    void testMultilineStringInterpolation() {
+        rewriteRun(
+          groovy(
+            """
+                String name = \"\"\"
+                    ${foo}
+                \"\"\"
+                """
+          )
+        );
+    }
+
 }

